### ISSUE TITLE
Support error codes 115 and 116 in is_coupon_valid()

### ISSUE THREAD

**Issue by Dekadinious** [woothemes/woocommerce#31171]
**Is your feature request related to a problem? Please describe.**
I am in the situation that a payment plugin puts orders in pending payment when they are processing the payment. If the payment fails, the orders are left there until cancelled by the core hold stock functionality.

The problem is that this holds coupons for the order. If we have coupons that can be used only once per customer, they will not be able to place their order. They will get an error message stating the coupon is already used. But from the customers standpoint, they have not used it.

Today we have solved it by listening for error code 115 and 116 in the coupon error filter. If they are present, we cancel all pending orders for that customer and change the error message to ask them to place the order again.

I feel this is not a good solution. Also, many customers don't understand it. It creates friction. I would like to be able to cancel the order before WooCommerce core checks if the coupon exist on another order for that customer. I therefore found is_coupon_valid() is the WC_Discounts class. Sadly, it doesn't handle error codes 115 or 116.

Could this be added so I can solve the above problem?

**Describe the solution you'd like**
Extend is_coupon_valid() to also check for errors 115 and 116.

**Describe alternatives you've considered**
Using other means to catch the error early. I am note sure what or how yet.

**Comment by juliaamosova** [comment]
Hi @Dekadinious,

Thank you for opening the issue! It requires further feedback from the WooCommerce Core team. I am adding the `needs developer feedback` label to this issue so that the Core team could take a look. 

Please note it may take a few days for them to get to this issue. Thank you for your patience.

**Comment by barryhughes** [comment]
> I therefore found is_coupon_valid() is the WC_Discounts class. Sadly, it doesn't handle error codes 115 or 116.

In my own testing, where a coupon is limited to one use per customer and where it has been applied to an order that is _pending payment,_ testing it with `WP_Discount::is_coupon_valid()` will return `(bool) false` as expected (assuming the current user is the same user associated with the order).

So, I'm not quite clear what change you are seeking here. Can you clarify, perhaps with an example of how you see this method (or any new method) behaving?

**Comment by Dekadinious** [comment]
@barryhughes Thanks for answering. It is not working as expected at least. Here is a minimal example:

```php
    function minimal_example()
    {
    error_log("1");

    $code = 'testcoupon';

    if (in_array($code, WC()->cart->get_applied_coupons())) {
        //Already applied
        return;
    }

    error_log("2");

    $coupon = new WC_Coupon($code);

    $discounts = new WC_Discounts();

    if ($discounts->is_coupon_valid($coupon) !== true) {
        error_log("3");

        $sessionCustomer = WC()->session->get('customer');

        $args = [
            'status' => ['wc-pending'],
            'customer' => sanitize_email($sessionCustomer['email']),
        ];

        $orders = wc_get_orders($args);

        if (empty($orders)) {
            error_log("4");
            return;
        }

        foreach ($orders as $order) {
            $order->set_status('wc-cancelled');
            $order->save();
        }
    }

    //Remove all coupons if other coupons are added
    if (!empty(WC()->cart->get_applied_coupons())) {
        WC()->cart->remove_coupons();
    }

    WC()->cart->apply_coupon($code);
    
    }
    add_action('woocommerce_before_checkout_form', 'minimal_example', 5);
```

This code will try to cancel the orders for the customer in pending payment if the coupon is not valid when the checkout is displayed. This is just a minimal example that tries to force adding of the coupon on every load of the cart.

This is a simple test procedure:

1. Create a coupon testcoupon with a maximum number of uses per customer of 1.
2. Add products to the cart and go to the checkout. The coupon should now be added.
3. Initiate checkout using an external payment method or another method that is configured to fail. An order will now be created in wc-pending.
4. Go back to the checkout from the failed payment. You will now trigger 'woocommerce_before_checkout_form'.

Expected behavior when reaching step 4 is that the log will contain 1, 2, and 3. The pending order should be canceled, and the coupon should be added so that the checkout is ready for the customer to just try again by hitting the button for checking out. Maybe they did not have funds on their card or have changed their payment method. The order should go straight through.

The actual behavior is that the log only contains 1 and 2. is_coupon_valid, therefore, returns true, even though there is an order in wc-pending blocking the usage of the coupon at this point. No orders are canceled. Checkout can not be done because of the blocking order in wc-pending. Refreshing the checkout page here gives an error message that something is wrong with the products in the cart. Trying the checkout again without refresh gives "Checkout Error". The customer is forced back to the cart, then back to the checkout whereby the log now triggers for 1, 2, and 3. The order is canceled and the customer can now place their order.

I am trying to avoid all of these hoops for the customer. If a payment fails with a coupon that can only be used once, the customer should be able to retry payment immediately from the checkout page.

PS: Sorry, can't get this code block to behave.

**Comment by barryhughes** [comment]
Understood.

I think the reason your approach is failing relates in some ways [to this other issue](https://github.com/woocommerce/woocommerce/issues/26289) (because, for a guest, the required check is deferred until very late in the checkout process, whereas you are expecting it to occur earlier).

```php
if ( ! is_user_logged_in() ) {
	// ...assuming you have constructed the coupon object and have the customer email from the session already...
	$was_applied = (bool) $coupon->get_data_store()->get_usage_by_email( $coupon, $customer_email );
	// ...logic to cancel the relevant orders...
}
```

You could potentially take advantage of logic like the above, when dealing with guests (and it looks to me as if your approach works for authenticated customers).

With that all noted ... I would caution there are presumably cases where there may be a real and valid delay before payment completes and where you might not wish this logic to run—but I'll leave it to you to factor that in.

I hope that helps: otherwise, this is essentially a custom dev question rather than a bug in itself so I'm going to close the issue.

**Comment by Dekadinious** [comment]
@barryhughes I am logged in while doing my testing. I am going to try your solution regardless and see how I do.

But would you say it makes for a good customer experience to be 100 % completely blocked from ordering through a single use coupon if the payment method fails and you have to try again? You are completely blocked until the order cancels itself or you get help from customer service. I can assure you that it happens a lot more often than you might think and that our customer service get's completely swamped when we have campaigns with single use coupons that get a fair amount of traffic.

**Comment by barryhughes** [comment]
> I am logged in while doing my testing.

Interesting. Then there must be some difference in how we're testing.

For me, though, when logged in as a customer and when a coupon with a use limit of 1-per-customer has already been applied to another order (which in this case is still pending payment), `$discounts->is_coupon_valid( $coupon )` returns false.

> But would you say it makes for a good customer experience to be 100 % completely blocked from ordering through a single use coupon if the payment method fails and you have to try again?

No, I would not.

I think a case could definitely be made to introduce functionality that handles this gracefully after payment fails *and where we know payment has failed* (ie, the order status is updated to 'failed') but that isn't the scenario we're dealing with here, as you've described it, because the order status remains in pending payment.

It's far less obvious to me that we should automatically assume payment has failed in that scenario. What payment gateway are you using?

**Comment by Dekadinious** [comment]
@barryhughes That's actually a good point. I hadn't considered that the payment gateway perhaps handled failed payments wrong. I have mentioned this issue to them, but I was assured that they handled things correctly. It's the Klarna plugin from Krokedil. I do not think it moves orders to a failed status when they fail, because they are just left in pending until canceled.

**Comment by barryhughes** [comment]
> I do not think it moves orders to a failed status when they fail, because they are just left in pending until canceled.

OK, well it is perhaps worth noting that various other gateways (Stripe is a good example) will generally mark an order as having failed near-instantly, if for example the card is invalid. However, what makes sense in this scenario will vary: perhaps the payment service does not provide the information they need to do anything else.

In any case, at least by default, the coupon will still not be 'released' until the order is actually cancelled which, as I think you noted, does not happen right away even if the order *is* marked as having failed—but it would at least be a lot easier to apply custom logic in that case: otherwise, your coupon usage limits could more easily be abused.

**Comment by Dekadinious** [comment]
Thank you for the explanation, I will see if they are able to get any sort of failed callback and can use the failed status in that case. Otherwise I will try your solution. Have a good one!
